faq/devcuda.inc

```diff
@@ -0,0 +1,50 @@
+<?php
+
+/////////////////////////////////////////////////////////////////////////
+
+$q[] = "What are some guidelines for developing CUDA-aware Open MPI applications?";
+
+$anchor[] = "mpi-cuda-dev";
+
+$a[] = "Developing CUDA-aware applications is a complex topic, and beyond
+the scope of this document. CUDA-aware applications often have to take 
+machine-specific considerations into account, including the number of GPUs
+installed on each node and how the GPUs are connected to the CPUs and to each
+other. Often, when using a particular transport layer (such as OPA/PSM2)
+there will be run-time decisions to make about which CPU cores will be used
+with which GPUs.
+
+A good place to start is the <a href=\"https://docs.nvidia.com/cuda/\">nVidia CUDA Toolkit Documentation</a>, including the <a href=\"https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html\">Programming Guide</a>
+and the <a href=\"https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html\">Best Practices Guide</a>.
+For examples of how to write CUDA-aware MPI applications, the <a href=\"https://github.com/NVIDIA-developer-blog/code-samples/tree/master/posts/cuda-aware-mpi-example\">nVidia developers blog</a>
+offers examples and the <a href=\"http://mvapich.cse.ohio-state.edu/benchmarks/\">OSU
+Micro-Benchmarks</a> offer an excellent example of how to write CUDA-aware MPI
+applications.";
+/////////////////////////////////////////////////////////////////////////
+
+$q[] = "What are some guidelines for using CUDA and Open MPI with Omni-Path?";
+
+$anchor[] = "mpi-cuda-dev-opa";
+
+$a[] = "When developing CUDA-aware Open MPI applications for OPA-based
+fabrics, the PSM2 transport is preferred and a CUDA-aware version of PSM2
+is provided with all versions of the Intel Omni-Path IFS software suite.
+
+The PSM2 library provides a number of settings that will govern how it
+will interact with CUDA, including <b>PSM2_CUDA</b> and <b>PSM2_GPUDIRECT</b>,
+which should be set in the environment before MPI_Init() is called. For
+example:
+
+<geshi bash>
+shell$ mpirun -x PSM2_CUDA=1 -x PSM2_GPUDIRECT=1 --mca mtl psm2 mpi_hello
+</geshi>
+
+In addition, each rank of the application should select a specific GPU card to
+use before calling MPI_Init(), by calling cudaSetDevice() or similar.
+The chosen GPU should be within the same NUMA node as the CPU the MPI rank is
+running on. (See the Linux taskset command and the sched_getaffinity and 
```

Comment:
I don't think recommending the taskset command here is a good idea.  For one thing, `hwloc` is an Open MPI project -- we should be recommending that.  For another thing, `mpirun` automatically does binding.

Reply:
Good point, using different MPIs all the time, I tend to forget that. I'll make the changes.